File: Dopamine.CollectionModule/ViewModels/CollectionGenresViewModel.vb
Imports System.Collections.ObjectModel
Imports Dopamine.Common.Presentation.Interfaces
Imports Dopamine.Common.Presentation.Utils
Imports Dopamine.Common.Presentation.ViewModels
Imports Dopamine.Common.Services.Metadata
Imports Dopamine.Common.Services.Playback
Imports Dopamine.Core.Base
Imports Dopamine.Core.Database.Entities
Imports Dopamine.Core.Database.Repositories.Interfaces
Imports Dopamine.Core.Logging
Imports Dopamine.Core.Prism
Imports Dopamine.Core.Settings
Imports Dopamine.Core.Utils
Imports Dopamine.Core.Helpers
Imports Microsoft.Practices.Prism.Commands
Imports Microsoft.Practices.Prism.PubSubEvents
Imports Dopamine.Core.Database

Namespace ViewModels
    Public Class CollectionGenresViewModel
        Inherits CommonAlbumsViewModel
        Implements ISemanticZoomViewModel

#Region "Variables"
        ' Repositories
        Private mGenreRepository As IGenreRepository

        ' Lists
        Private mGenres As ObservableCollection(Of ISemanticZoomable)
        Private mGenresCvs As CollectionViewSource
        Private mSelectedGenres As IList(Of Genre)
        Private mGenresZoomSelectors As ObservableCollection(Of ISemanticZoomSelector)

        ' Flags
        Private mIsGenresZoomVisible As Boolean

        ' Other
        Private mGenresCount As Long
        Private mShellMouseUpToken As SubscriptionToken
        Private mLeftPaneWidthPercent As Double
        Private mRightPaneWidthPercent As Double
#End Region

#Region "Commands"
        Public Property AddGenresToPlaylistCommand As DelegateCommand(Of String)
        Public Property SelectedGenresCommand As DelegateCommand(Of Object)
        Public Property ShowGenresZoomCommand As DelegateCommand
        Public Property SemanticJumpCommand As DelegateCommand Implements ISemanticZoomViewModel.SemanticJumpCommand
        Public Property AddGenresToNowPlayingCommand As DelegateCommand
#End Region

#Region "Properties"
        Public Property LeftPaneWidthPercent() As Double
            Get
                Return mLeftPaneWidthPercent
            End Get
            Set(ByVal value As Double)
                SetProperty(Of Double)(mLeftPaneWidthPercent, value)
                XmlSettingsClient.Instance.Set(Of Integer)("ColumnWidths", "GenresLeftPaneWidthPercent", CInt(value))
            End Set
        End Property

        Public Property RightPaneWidthPercent() As Double
            Get
                Return mRightPaneWidthPercent
            End Get
            Set(ByVal value As Double)
                SetProperty(Of Double)(mRightPaneWidthPercent, value)
                XmlSettingsClient.Instance.Set(Of Integer)("ColumnWidths", "GenresRightPaneWidthPercent", CInt(value))
            End Set
        End Property

        Public Property Genres() As ObservableCollection(Of ISemanticZoomable) Implements ISemanticZoomViewModel.SemanticZoomables
            Get
                Return mGenres
            End Get
            Set(ByVal value As ObservableCollection(Of ISemanticZoomable))
                SetProperty(Of ObservableCollection(Of ISemanticZoomable))(mGenres, value)
            End Set
        End Property

        Public Property GenresCvs() As CollectionViewSource
            Get
                Return mGenresCvs
            End Get
            Set(ByVal value As CollectionViewSource)
                SetProperty(Of CollectionViewSource)(mGenresCvs, value)
            End Set
        End Property

        Public Property SelectedGenres() As IList(Of Genre)
            Get
                Return mSelectedGenres
            End Get
            Set(ByVal value As IList(Of Genre))
                SetProperty(Of IList(Of Genre))(mSelectedGenres, value)
            End Set
        End Property

        Public Property GenresCount() As Long
            Get
                Return mGenresCount
            End Get
            Set(ByVal value As Long)
                SetProperty(Of Long)(mGenresCount, value)
            End Set
        End Property

        Public Property IsGenresZoomVisible() As Boolean
            Get
                Return mIsGenresZoomVisible
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsGenresZoomVisible, value)
            End Set
        End Property

        Public Property GenresZoomSelectors() As ObservableCollection(Of ISemanticZoomSelector) Implements ISemanticZoomViewModel.SemanticZoomSelectors
            Get
                Return mGenresZoomSelectors
            End Get
            Set(ByVal value As ObservableCollection(Of ISemanticZoomSelector))
                SetProperty(Of ObservableCollection(Of ISemanticZoomSelector))(mGenresZoomSelectors, value)
            End Set
        End Property

        Public Overrides ReadOnly Property CanOrderByAlbum As Boolean
            Get
                Return (Me.SelectedGenres IsNot Nothing AndAlso
                       Me.SelectedGenres.Count > 0) Or
                       (Me.SelectedAlbums IsNot Nothing AndAlso
                       Me.SelectedAlbums.Count > 0)
            End Get
        End Property
#End Region

#Region "Construction"
        Public Sub New(iGenreRepository As IGenreRepository)

            MyBase.New()

            ' Repositories
            mGenreRepository = iGenreRepository

            ' Commands
            Me.ToggleTrackOrderCommand = New DelegateCommand(Async Sub() Await Me.ToggleTrackOrderAsync)
            Me.ToggleAlbumOrderCommand = New DelegateCommand(Async Sub() Await Me.ToggleAlbumOrderAsync)
            Me.RemoveSelectedTracksCommand = New DelegateCommand(Async Sub() Await Me.RemoveTracksFromCollectionAsync(Me.SelectedTracks), Function() Not Me.IsIndexing)
            Me.AddGenresToPlaylistCommand = New DelegateCommand(Of String)(Async Sub(iPlaylistName) Await Me.AddGenresToPlaylistAsync(Me.SelectedGenres, iPlaylistName))
            Me.SelectedGenresCommand = New DelegateCommand(Of Object)(Async Sub(iParameter) Await Me.SelectedGenresHandlerAsync(iParameter))
            Me.ShowGenresZoomCommand = New DelegateCommand(Async Sub() Await Me.ShowSemanticZoomAsync())
            Me.SemanticJumpCommand = New DelegateCommand(Sub() Me.IsGenresZoomVisible = False)
            Me.AddGenresToNowPlayingCommand = New DelegateCommand(Async Sub() Await Me.AddGenresToNowPlayingAsync(Me.SelectedGenres))

            ' Events
            Me.eventAggregator.GetEvent(Of SettingEnableRatingChanged)().Subscribe(Async Sub(iEnableRating)
                                                                                       Me.EnableRating = iEnableRating
                                                                                       Me.SetTrackOrder("GenresTrackOrder")
                                                                                       Await Me.GetTracksAsync(Nothing, Me.SelectedGenres, Me.SelectedAlbums, Me.TrackOrder)
                                                                                   End Sub)

            Me.eventAggregator.GetEvent(Of SettingUseStarRatingChanged)().Subscribe(Sub(iUseStarRating) Me.UseStarRating = iUseStarRating)

            ' MetadataService
            AddHandler Me.metadataService.MetadataChanged, AddressOf MetadataChangedHandlerAsync

            ' IndexingService
            AddHandler Me.indexingService.RefreshArtwork, Async Sub() Await Me.collectionService.RefreshArtworkAsync(Me.Albums, Me.Tracks)

            ' Set the initial AlbumOrder
            Me.AlbumOrder = CType(XmlSettingsClient.Instance.Get(Of Integer)("Ordering", "GenresAlbumOrder"), AlbumOrder)

            ' Set the initial TrackOrder
            Me.SetTrackOrder("GenresTrackOrder")

            ' Subscribe to Events and Commands on creation
            Me.Subscribe()

            ' Set width of the panels
            Me.LeftPaneWidthPercent = XmlSettingsClient.Instance.Get(Of Integer)("ColumnWidths", "GenresLeftPaneWidthPercent")
            Me.RightPaneWidthPercent = XmlSettingsClient.Instance.Get(Of Integer)("ColumnWidths", "GenresRightPaneWidthPercent")

            ' Cover size
            Me.SetCoversizeasync(CType(XmlSettingsClient.Instance.Get(Of Integer)("CoverSizes", "GenresCoverSize"), CoverSizeType))
        End Sub
#End Region

#Region "ISemanticZoomViewModel"
        Public Async Function ShowSemanticZoomAsync() As Task Implements ISemanticZoomViewModel.ShowSemanticZoomAsync

            Me.GenresZoomSelectors = Await SemanticZoomUtils.UpdateSemanticZoomSelectors(Me.GenresCvs.View)

            Me.IsGenresZoomVisible = True
        End Function

        Public Sub HideSemanticZoom() Implements ISemanticZoomViewModel.HideSemanticZoom
            Me.IsGenresZoomVisible = False
        End Sub

        Public Sub UpdateSemanticZoomHeaders() Implements ISemanticZoomViewModel.UpdateSemanticZoomHeaders

            Dim previousHeader As String = String.Empty

            For Each gvm As GenreViewModel In Me.GenresCvs.View

                If String.IsNullOrEmpty(previousHeader) OrElse Not gvm.Header.Equals(previousHeader) Then

                    previousHeader = gvm.Header
                    gvm.IsHeader = True
                End If
            Next
        End Sub
#End Region

#Region "Private"
        Private Async Sub MetadataChangedHandlerAsync(e As MetadataChangedEventArgs)

            If e.IsAlbumArtworkMetadataChanged Then
                Await Me.collectionService.RefreshArtworkAsync(Me.Albums, Me.Tracks)
            End If

            If e.IsGenreMetadataChanged Then
                Await Me.GetGenresAsync()
            End If

            If e.IsGenreMetadataChanged Or e.IsAlbumTitleMetadataChanged Or e.IsAlbumArtistMetadataChanged Or e.IsAlbumYearMetadataChanged Then
                Await Me.GetAlbumsAsync(Nothing, Me.SelectedGenres, Me.AlbumOrder)
            End If

            If e.IsGenreMetadataChanged Or e.IsAlbumTitleMetadataChanged Or e.IsAlbumArtistMetadataChanged Or e.IsTrackMetadataChanged Then
                Await Me.GetTracksAsync(Nothing, Me.SelectedGenres, Me.SelectedAlbums, Me.TrackOrder)
            End If
        End Sub

        Private Async Function GetGenresAsync() As Task

            Try
                ' Get Genres from database
                Dim genres As List(Of Genre) = Await mGenreRepository.GetGenresAsync()

                ' Create new ObservableCollection
                Dim genreViewModels As New ObservableCollection(Of GenreViewModel)

                Await Task.Run(Sub()
                                   Dim tempGenreViewModels As New List(Of GenreViewModel)

                                   ' Workaround to make sure the "#" GroupHeader is shown at the top of the list
                                   tempGenreViewModels.AddRange(genres.Select(Function(gen) New GenreViewModel With {.Genre = gen, .IsHeader = False}).Where(Function(gvm) gvm.Header.Equals("#")))
                                   tempGenreViewModels.AddRange(genres.Select(Function(gen) New GenreViewModel With {.Genre = gen, .IsHeader = False}).Where(Function(gvm) Not gvm.Header.Equals("#")))

                                   For Each gvm As GenreViewModel In tempGenreViewModels
                                       genreViewModels.Add(gvm)
                                   Next
                               End Sub)

                ' Unbind to improve UI performance
                If Me.GenresCvs IsNot Nothing Then RemoveHandler Me.GenresCvs.Filter, New FilterEventHandler(AddressOf GenresCvs_Filter)
                Me.Genres = Nothing
                Me.GenresCvs = Nothing

                ' Populate ObservableCollection
                Me.Genres = New ObservableCollection(Of ISemanticZoomable)(genreViewModels)
            Catch ex As Exception
                LogClient.Instance.Logger.Error("An error occurred while getting Genres. Exception: {0}", ex.Message)

                ' Failed getting Genres. Create empty ObservableCollection.
                Me.Genres = New ObservableCollection(Of ISemanticZoomable)
            End Try

            ' Populate CollectionViewSource
            Me.GenresCvs = New CollectionViewSource With {.Source = Me.Genres}
            AddHandler Me.GenresCvs.Filter, New FilterEventHandler(AddressOf GenresCvs_Filter)

            ' Update count
            Me.GenresCount = Me.GenresCvs.View.Cast(Of ISemanticZoomable)().Count

            ' Update Semantic Zoom Headers
            Me.UpdateSemanticZoomHeaders()
        End Function

        Private Async Function SelectedGenresHandlerAsync(iParameter As Object) As Task

            If iParameter IsNot Nothing Then

                Me.SelectedGenres = New List(Of Genre)

                For Each item As GenreViewModel In CType(iParameter, IList)
                    Me.SelectedGenres.Add(item.Genre)
                Next
            End If

            ' Don't reload the lists when updating Metadata. MetadataChangedHandlerAsync handles that.
            If Me.metadataService.IsUpdatingDatabaseMetadata Then Return

            Await Me.GetAlbumsAsync(Nothing, Me.SelectedGenres, CType(XmlSettingsClient.Instance.Get(Of Integer)("Ordering", "GenresAlbumOrder"), AlbumOrder))
            Me.SetTrackOrder("GenresTrackOrder")
            Await Me.GetTracksAsync(Nothing, Me.SelectedGenres, Me.SelectedAlbums, Me.TrackOrder)
        End Function

        Private Async Function AddGenresToPlaylistAsync(iGenres As IList(Of Genre), iPlaylistName As String) As Task

            Dim result As AddToPlaylistResult = Await Me.collectionService.AddGenresToPlaylistAsync(iGenres, iPlaylistName)

            If Not result.IsSuccess Then

                Me.dialogService.ShowNotification(&HE711,
                                                16,
                                                ResourceUtils.GetStringResource("Language_Error"),
                                                Replace(ResourceUtils.GetStringResource("Language_Error_Adding_Genres_To_Playlist"), "%playlistname%", """" & iPlaylistName & """"),
                                                ResourceUtils.GetStringResource("Language_Ok"),
                                                True,
                                                ResourceUtils.GetStringResource("Language_Log_File"))
            End If
        End Function

        Private Async Function AddGenresToNowPlayingAsync(iGenres As IList(Of Genre)) As Task

            Dim result As AddToQueueResult = Await Me.playbackService.AddToQueue(iGenres)

            If Not result.IsSuccess Then

                Me.dialogService.ShowNotification(&HE711,
                                                16,
                                                ResourceUtils.GetStringResource("Language_Error"),
                                                ResourceUtils.GetStringResource("Language_Error_Adding_Genres_To_Now_Playing"),
                                                ResourceUtils.GetStringResource("Language_Ok"),
                                                True,
                                                ResourceUtils.GetStringResource("Language_Log_File"))
            End If
        End Function

        Private Sub GenresCvs_Filter(sender As Object, e As FilterEventArgs)

            Dim gvm As GenreViewModel = TryCast(e.Item, GenreViewModel)

            e.Accepted = Dopamine.Core.Database.Utils.FilterGenres(gvm.Genre, Me.searchService.SearchText)
        End Sub
#End Region

#Region "Protected"
        Protected Async Function ToggleTrackOrderAsync() As Task

            MyBase.ToggleTrackOrder()

            XmlSettingsClient.Instance.Set(Of Integer)("Ordering", "GenresTrackOrder", Me.TrackOrder)
            Await Me.GetTracksCommonAsync(Me.Tracks.Select(Function(t) t.TrackInfo).ToList, Me.TrackOrder)
        End Function

        Protected Async Function ToggleAlbumOrderAsync() As Task

            MyBase.ToggleAlbumOrder()

            XmlSettingsClient.Instance.Set(Of Integer)("Ordering", "GenresAlbumOrder", Me.AlbumOrder)
            Await Me.GetAlbumsCommonAsync(Me.Albums.Select(Function(a) a.Album).ToList, Me.AlbumOrder)
        End Function
#End Region

#Region "Overrides"
        Protected Overrides Async Function SetCoversizeasync(iCoverSize As CoverSizeType) As Task

            Await MyBase.SetCoversizeAsync(iCoverSize)
            XmlSettingsClient.Instance.Set(Of Integer)("CoverSizes", "GenresCoverSize", iCoverSize)
        End Function

        Protected Overrides Async Function FillListsAsync() As Task

            Await Me.GetGenresAsync()
            Await Me.GetAlbumsAsync(Nothing, Nothing, Me.AlbumOrder)
            Await Me.GetTracksAsync(Nothing, Nothing, Nothing, Me.TrackOrder)
        End Function

        Protected Overrides Sub FilterLists()

            Application.Current.Dispatcher.Invoke(Sub()
                                                      ' Genres
                                                      If Me.GenresCvs IsNot Nothing Then
                                                          Me.GenresCvs.View.Refresh()
                                                          Me.GenresCount = Me.GenresCvs.View.Cast(Of ISemanticZoomable)().Count
                                                          Me.UpdateSemanticZoomHeaders()
                                                      End If
                                                  End Sub)

            MyBase.FilterLists()
        End Sub

        Protected Overrides Async Function SelectedAlbumsHandlerAsync(iParameter As Object) As Task

            Await MyBase.SelectedAlbumsHandlerAsync(iParameter)

            Me.SetTrackOrder("GenresTrackOrder")
            Await Me.GetTracksAsync(Nothing, Me.SelectedGenres, Me.SelectedAlbums, Me.TrackOrder)
        End Function

        Protected Overrides Sub Unsubscribe()

            ' Commands
            ApplicationCommands.RemoveSelectedTracksCommand.UnregisterCommand(Me.RemoveSelectedTracksCommand)
            ApplicationCommands.SemanticJumpCommand.UnregisterCommand(Me.SemanticJumpCommand)

            ' Events
            Me.eventAggregator.GetEvent(Of ShellMouseUp).Unsubscribe(mShellMouseUpToken)

            ' Other
            Me.IsGenresZoomVisible = False
        End Sub

        Protected Overrides Sub Subscribe()

            ' Prevents subscribing twice
            Me.Unsubscribe()

            ' Commands
            ApplicationCommands.RemoveSelectedTracksCommand.RegisterCommand(Me.RemoveSelectedTracksCommand)
            ApplicationCommands.SemanticJumpCommand.RegisterCommand(Me.SemanticJumpCommand)

            ' Events
            mShellMouseUpToken = Me.eventAggregator.GetEvent(Of ShellMouseUp).Subscribe(Sub() Me.IsGenresZoomVisible = False)
        End Sub

        Protected Overrides Sub RefreshLanguage()

            Me.UpdateAlbumOrderText(Me.AlbumOrder)
            Me.UpdateTrackOrderText(Me.TrackOrder)
        End Sub
#End Region
    End Class
End Namespace

File: Dopamine/ViewModels/FullPlayer/Collection/CollectionGenresViewModel.cs
using Digimezzo.Foundation.Core.Logging;
using Digimezzo.Foundation.Core.Settings;
using Digimezzo.Foundation.Core.Utils;
using Dopamine.Core.Base;
using Dopamine.Core.Prism;
using Dopamine.Data;
using Dopamine.Services.Collection;
using Dopamine.Services.Dialog;
using Dopamine.Services.Entities;
using Dopamine.Services.Indexing;
using Dopamine.Services.Playback;
using Dopamine.Services.Playlist;
using Dopamine.Services.Search;
using Dopamine.Services.Utils;
using Dopamine.ViewModels.Common.Base;
using Prism.Commands;
using Prism.Events;
using Prism.Ioc;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Data;

namespace Dopamine.ViewModels.FullPlayer.Collection
{
    public class CollectionGenresViewModel : AlbumsViewModelBase, ISemanticZoomViewModel
    {
        private ICollectionService collectionService;
        private IIndexingService indexingService;
        private IDialogService dialogService;
        private IPlaybackService playbackService;
        private IPlaylistService playlistService;
        private ISearchService searchService;
        private IEventAggregator eventAggregator;
        private ObservableCollection<ISemanticZoomable> genres;
        private CollectionViewSource genresCvs;
        private IList<string> selectedGenres;
        private ObservableCollection<ISemanticZoomSelector> genresZoomSelectors;
        private bool isGenresZoomVisible;
        private long genresCount;
        private double leftPaneWidthPercent;
        private double rightPaneWidthPercent;

        public DelegateCommand<string> AddGenresToPlaylistCommand { get; set; }

        public DelegateCommand<object> SelectedGenresCommand { get; set; }

        public DelegateCommand ShowGenresZoomCommand { get; set; }

        public DelegateCommand<string> SemanticJumpCommand { get; set; }

        public DelegateCommand AddGenresToNowPlayingCommand { get; set; }

        public DelegateCommand ShuffleSelectedGenresCommand { get; set; }

        public double LeftPaneWidthPercent
        {
            get { return this.leftPaneWidthPercent; }
            set
            {
                SetProperty<double>(ref this.leftPaneWidthPercent, value);
                SettingsClient.Set<int>("ColumnWidths", "GenresLeftPaneWidthPercent", Convert.ToInt32(value));
            }
        }

        public double RightPaneWidthPercent
        {
            get { return this.rightPaneWidthPercent; }
            set
            {
                SetProperty<double>(ref this.rightPaneWidthPercent, value);
                SettingsClient.Set<int>("ColumnWidths", "GenresRightPaneWidthPercent", Convert.ToInt32(value));
            }
        }

        public ObservableCollection<ISemanticZoomable> Genres
        {
            get { return this.genres; }
            set { SetProperty<ObservableCollection<ISemanticZoomable>>(ref this.genres, value); }
        }
        ObservableCollection<ISemanticZoomable> ISemanticZoomViewModel.SemanticZoomables
        {
            get { return Genres; }
            set { Genres = value; }
        }

        public CollectionViewSource GenresCvs
        {
            get { return this.genresCvs; }
            set { SetProperty<CollectionViewSource>(ref this.genresCvs, value); }
        }

        public IList<string> SelectedGenres
        {
            get { return this.selectedGenres; }
            set { SetProperty<IList<string>>(ref this.selectedGenres, value); }
        }

        public long GenresCount
        {
            get { return this.genresCount; }
            set { SetProperty<long>(ref this.genresCount, value); }
        }

        public bool IsGenresZoomVisible
        {
            get { return this.isGenresZoomVisible; }
            set { SetProperty<bool>(ref this.isGenresZoomVisible, value); }
        }

        public ObservableCollection<ISemanticZoomSelector> GenresZoomSelectors
        {
            get { return this.genresZoomSelectors; }
            set { SetProperty<ObservableCollection<ISemanticZoomSelector>>(ref this.genresZoomSelectors, value); }
        }
        ObservableCollection<ISemanticZoomSelector> ISemanticZoomViewModel.SemanticZoomSelectors
        {
            get { return GenresZoomSelectors; }
            set { GenresZoomSelectors = value; }
        }

        public CollectionGenresViewModel(IContainerProvider container) : base(container)
        {
            // Dependency injection
            this.collectionService = container.Resolve<ICollectionService>();
            this.dialogService = container.Resolve<IDialogService>();
            this.indexingService = container.Resolve<IIndexingService>();
            this.playbackService = container.Resolve<IPlaybackService>();
            this.playlistService = container.Resolve<IPlaylistService>();
            this.searchService = container.Resolve<ISearchService>();
            this.eventAggregator = container.Resolve<IEventAggregator>();

            // Commands
            this.ToggleTrackOrderCommand = new DelegateCommand(async () => await this.ToggleTrackOrderAsync());
            this.ToggleAlbumOrderCommand = new DelegateCommand(async () => await this.ToggleAlbumOrderAsync());
            this.RemoveSelectedTracksCommand = new DelegateCommand(async () => await this.RemoveTracksFromCollectionAsync(this.SelectedTracks), () => !this.IsIndexing);
            this.AddGenresToPlaylistCommand = new DelegateCommand<string>(async (playlistName) => await this.AddGenresToPlaylistAsync(this.SelectedGenres, playlistName));
            this.SelectedGenresCommand = new DelegateCommand<object>(async (parameter) => await this.SelectedGenresHandlerAsync(parameter));
            this.ShowGenresZoomCommand = new DelegateCommand(async () => await this.ShowSemanticZoomAsync());
            this.AddGenresToNowPlayingCommand = new DelegateCommand(async () => await this.AddGenresToNowPlayingAsync(this.SelectedGenres));
            this.ShuffleSelectedGenresCommand = new DelegateCommand(async () => await this.playbackService.EnqueueGenresAsync(this.SelectedGenres, true, false));

            this.SemanticJumpCommand = new DelegateCommand<string>((header) =>
            {
                this.HideSemanticZoom();
                this.eventAggregator.GetEvent<PerformSemanticJump>().Publish(new Tuple<string, string>("Genres", header));
            });

            // Settings
            SettingsClient.SettingChanged += async (_, e) =>
            {
                if (SettingsClient.IsSettingChanged(e, "Behaviour", "EnableRating"))
                {
                    this.EnableRating = (bool)e.Entry.Value;
                    this.SetTrackOrder("GenresTrackOrder");
                    await this.GetTracksAsync(null, this.SelectedGenres, this.SelectedAlbums, this.TrackOrder);
                }

                if (SettingsClient.IsSettingChanged(e, "Behaviour", "EnableLove"))
                {
                    this.EnableLove = (bool)e.Entry.Value;
                    this.SetTrackOrder("GenresTrackOrder");
                    await this.GetTracksAsync(null, this.SelectedGenres, this.SelectedAlbums, this.TrackOrder);
                }
            };

            // PubSub Events
            this.eventAggregator.GetEvent<ShellMouseUp>().Subscribe((_) => this.IsGenresZoomVisible = false);

            // Set the initial AlbumOrder
            this.AlbumOrder = (AlbumOrder)SettingsClient.Get<int>("Ordering", "GenresAlbumOrder");

            // Set the initial TrackOrder
            this.SetTrackOrder("GenresTrackOrder");

            // Set width of the panels
            this.LeftPaneWidthPercent = SettingsClient.Get<int>("ColumnWidths", "GenresLeftPaneWidthPercent");
            this.RightPaneWidthPercent = SettingsClient.Get<int>("ColumnWidths", "GenresRightPaneWidthPercent");

            // Cover size
            this.SetCoversizeAsync((CoverSizeType)SettingsClient.Get<int>("CoverSizes", "GenresCoverSize"));
        }

        public async Task ShowSemanticZoomAsync()
        {
            this.GenresZoomSelectors = await SemanticZoomUtils.UpdateSemanticZoomSelectors(this.GenresCvs.View);

            this.IsGenresZoomVisible = true;
        }

        public void HideSemanticZoom()
        {
            this.IsGenresZoomVisible = false;
        }

        public void UpdateSemanticZoomHeaders()
        {
            string previousHeader = string.Empty;

            foreach (GenreViewModel gvm in this.GenresCvs.View)
            {
                if (string.IsNullOrEmpty(previousHeader) || !gvm.Header.Equals(previousHeader))
                {
                    previousHeader = gvm.Header;
                    gvm.IsHeader = true;
                }
                else
                {
                    gvm.IsHeader = false;
                }
            }
        }

        private void ClearGenres()
        {
            Application.Current.Dispatcher.Invoke(() =>
            {
                if (this.GenresCvs != null)
                {
                    this.GenresCvs.Filter -= new FilterEventHandler(GenresCvs_Filter);
                }

                this.GenresCvs = null;
            });

            this.Genres = null;
        }

        private async Task GetGenresAsync()
        {
            try
            {
                // Get the genres
                var genreViewModels = new ObservableCollection<GenreViewModel>(await this.collectionService.GetAllGenresAsync());

                // Unbind to improve UI performance
                this.ClearGenres();

                // Populate ObservableCollection
                this.Genres = new ObservableCollection<ISemanticZoomable>(genreViewModels);
            }
            catch (Exception ex)
            {
                LogClient.Error("An error occurred while getting Genres. Exception: {0}", ex.Message);

                // Failed getting Genres. Create empty ObservableCollection.
                this.Genres = new ObservableCollection<ISemanticZoomable>();
            }

            Application.Current.Dispatcher.Invoke(() =>
            {
                // Populate CollectionViewSource
                this.GenresCvs = new CollectionViewSource { Source = this.Genres };
                this.GenresCvs.Filter += new FilterEventHandler(GenresCvs_Filter);

                // Update count
                this.GenresCount = this.GenresCvs.View.Cast<ISemanticZoomable>().Count();
            });

            // Update Semantic Zoom Headers
            this.UpdateSemanticZoomHeaders();
        }

        private async Task SelectedGenresHandlerAsync(object parameter)
        {
            if (parameter != null)
            {
                this.SelectedGenres = new List<string>();

                foreach (GenreViewModel item in (IList)parameter)
                {
                    this.SelectedGenres.Add(item.GenreName);
                }
            }

            await this.GetGenreAlbumsAsync(this.SelectedGenres, this.AlbumOrder);
            this.SetTrackOrder("GenresTrackOrder");
            await this.GetTracksAsync(null, this.SelectedGenres, this.SelectedAlbums, this.TrackOrder);
        }

        private async Task AddGenresToPlaylistAsync(IList<string> genres, string playlistName)
        {
            CreateNewPlaylistResult createPlaylistResult = CreateNewPlaylistResult.Success; // Default Success

            // If no playlist is provided, first create one.
            if (playlistName == null)
            {
                var responseText = ResourceUtils.GetString("Language_New_Playlist");

                if (this.dialogService.ShowInputDialog(
                    0xea37,
                    16,
                    ResourceUtils.GetString("Language_New_Playlist"),
                    ResourceUtils.GetString("Language_Enter_Name_For_Playlist"),
                    ResourceUtils.GetString("Language_Ok"),
                    ResourceUtils.GetString("Language_Cancel"),
                    ref responseText))
                {
                    playlistName = responseText;
                    createPlaylistResult = await this.playlistService.CreateNewPlaylistAsync(new EditablePlaylistViewModel(playlistName, PlaylistType.Static));
                }
            }

            // If playlist name is still null, the user clicked cancel on the previous dialog. Stop here.
            if (playlistName == null)
            {
                return;
            }

            // Verify if the playlist was created
            switch (createPlaylistResult)
            {
                case CreateNewPlaylistResult.Success:
                case CreateNewPlaylistResult.Duplicate:
                    // Add items to playlist
                    AddTracksToPlaylistResult addTracksResult = await this.playlistService.AddGenresToStaticPlaylistAsync(genres, playlistName);

                    if (addTracksResult == AddTracksToPlaylistResult.Error)
                    {
                        this.dialogService.ShowNotification(0xe711, 16, ResourceUtils.GetString("Language_Error"), ResourceUtils.GetString("Language_Error_Adding_Songs_To_Playlist").Replace("{playlistname}", "\"" + playlistName + "\""), ResourceUtils.GetString("Language_Ok"), true, ResourceUtils.GetString("Language_Log_File"));
                    }
                    break;
                case CreateNewPlaylistResult.Error:
                    this.dialogService.ShowNotification(
                        0xe711,
                        16,
                        ResourceUtils.GetString("Language_Error"),
                        ResourceUtils.GetString("Language_Error_Adding_Playlist"),
                        ResourceUtils.GetString("Language_Ok"),
                        true,
                        ResourceUtils.GetString("Language_Log_File"));
                    break;
                case CreateNewPlaylistResult.Blank:
                    this.dialogService.ShowNotification(
                        0xe711,
                        16,
                        ResourceUtils.GetString("Language_Error"),
                        ResourceUtils.GetString("Language_Provide_Playlist_Name"),
                        ResourceUtils.GetString("Language_Ok"),
                        false,
                        string.Empty);
                    break;
                default:
                    // Never happens
                    break;
            }
        }

        private async Task AddGenresToNowPlayingAsync(IList<string> genres)
        {
            EnqueueResult result = await this.playbackService.AddGenresToQueueAsync(genres);

            if (!result.IsSuccess)
            {
                this.dialogService.ShowNotification(0xe711, 16, ResourceUtils.GetString("Language_Error"), ResourceUtils.GetString("Language_Error_Adding_Genres_To_Now_Playing"), ResourceUtils.GetString("Language_Ok"), true, ResourceUtils.GetString("Language_Log_File"));
            }
        }

        private void GenresCvs_Filter(object sender, FilterEventArgs e)
        {
            GenreViewModel gvm = e.Item as GenreViewModel;

            e.Accepted = Services.Utils.EntityUtils.FilterGenres(gvm, this.searchService.SearchText);
        }

        private async Task ToggleTrackOrderAsync()
        {
            base.ToggleTrackOrder();

            SettingsClient.Set<int>("Ordering", "GenresTrackOrder", (int)this.TrackOrder);
            await this.GetTracksCommonAsync(this.Tracks, this.TrackOrder);
        }

        private async Task ToggleAlbumOrderAsync()
        {
            base.ToggleAlbumOrder();

            SettingsClient.Set<int>("Ordering", "GenresAlbumOrder", (int)this.AlbumOrder);
            await this.GetAlbumsCommonAsync(this.Albums, this.AlbumOrder);
        }

        protected async override Task SetCoversizeAsync(CoverSizeType coverSize)
        {
            await base.SetCoversizeAsync(coverSize);
            SettingsClient.Set<int>("CoverSizes", "GenresCoverSize", (int)coverSize);
        }

        protected async override Task FillListsAsync()
        {
            await this.GetGenresAsync();
            await this.GetGenreAlbumsAsync(this.SelectedGenres, this.AlbumOrder);
            await this.GetTracksAsync(null, this.SelectedGenres, this.SelectedAlbums, this.TrackOrder);
        }

        protected async override Task EmptyListsAsync()
        {
            this.ClearGenres();
            this.ClearAlbums();
            this.ClearTracks();
        }

        protected override void FilterLists()
        {
            Application.Current.Dispatcher.Invoke(() =>
            {
                // Genres
                if (this.GenresCvs != null)
                {
                    this.GenresCvs.View.Refresh();
                    this.GenresCount = this.GenresCvs.View.Cast<ISemanticZoomable>().Count();
                    this.UpdateSemanticZoomHeaders();
                }
            });

            base.FilterLists();
        }

        protected async override Task SelectedAlbumsHandlerAsync(object parameter)
        {
            await base.SelectedAlbumsHandlerAsync(parameter);

            this.SetTrackOrder("GenresTrackOrder");
            await this.GetTracksAsync(null, this.SelectedGenres, this.SelectedAlbums, this.TrackOrder);
        }

        protected override void RefreshLanguage()
        {
            this.UpdateAlbumOrderText(this.AlbumOrder);
            this.UpdateTrackOrderText(this.TrackOrder);
            base.RefreshLanguage();
        }
    }
}
